[RPC] Synchronous, typed interface for Python client

m-yac opened the pull request:
This PR adds a synchronous and type-annotated interface for the python Cryptol RPC bindings. The advantages of such an interface are:
- No more calls to `.result()`, since everything is synchronous
- `mypy` can give useful feedback about the results of running RPC commands, since everything is type-annotated (for example, the return type of `check` is annotated to be a `CheckResult`)

All the tests have been updated to use the new interface, though the old interface still exists and is functionally unchanged ~~(aside from the one exception noted below)~~.

This PR also makes the following two changes that affect both the old and new interfaces:
- The `CheckReport` class now has `__bool__` method which returns the value of its `success` field (e.g. so `self.assert(c.check(...).success)` can be replaced with just `self.assert(c.check(...))`).
- ~~Instead of returning a `Union[bool, List[CryptolPython], OfflineSmt]`, the `sat`, `prove`, and `safe` methods now return a `Union[SmtResult, OfflineSmtQuery]`, where `SmtResult` has the subclasses `SmtUnsatResult`, `SmtSatResult`, and `SmtInvalidResult`. The `SmtUnsatResult` class has a `__bool__` method which corresponds to the old `bool` case and the `SmtSatResult` and `SmtInvalidResult` classes are subclasses of `List[CryptolPython]`. Thus, this change will only break code which explicitly checks whether the result of `sat`/`prove`/`safe` is a `bool` (e.g. `isinstance(c.sat(...).result(), bool)`) but not code which just uses it as a truthy value (e.g. `if c.sat(...).result(): ...`).~~

**Question**: I called the type of the output of `from_cryptol_arg` "CryptolPython", but I feel like this could potentially be confusing. Thoughts on a better name?

**NB**: The definition of `CryptolPython` really should be:
```
CryptolPython = Union[bool, int, BV, Tuple[CryptolPython], List[CryptolPython], Dict[str, CryptolPython], OpaqueValue]
```
but `mypy` doesn't yet support recursive types.

pnwamk commented on cryptol-remote-api/python/cryptol/commands.py:
```diff
@@ -161,14 +183,62 @@ def __init__(self, connection : HasProtocolState, expr : Any) -> None:
     SAT   = 'sat'
 
-class CryptolProveSat(argo.Command):
+class SmtResult(ABC):
```
I'm wondering if it would be clearer for more users (me included) if `sat`, `prove`, and `safe` returned types more specialized for the question they are answering, similar to what the Cryptol REPL reports. E.g.,

`:prove` at the REPL reports counterexample or QED:
```
Cryptol> :prove 1+1 == 0
Counterexample
(1 + 1 == 0) = False
(Total Elapsed Time: 0.009s, using "Z3")

Cryptol> :prove 1+1 == 2
Q.E.D.
(Total Elapsed Time: 0.009s, using "Z3")
```

so maybe `prove` in python returns `Union[CounterExample, QED]`?

`:sat` simply tells you whether or not the expression is satisfiable:
```
Cryptol> :sat \(x:[8]) -> 1+1 == x
Satisfiable
(\(x : [8]) -> 1 + 1 == x) 0x02 = True
(Total Elapsed Time: 0.010s, using "Z3")

Cryptol> :sat \(x:[8]) -> 1+x == 2+x
Unsatisfiable
(Total Elapsed Time: 0.010s, using "Z3")
Cryptol> 
```

perhaps `sat` in python returns `Union[Satisfiable, Unsatisfiable]`?

`:safe` reports either that indeed it was safe or it gives you a specific counter example:
```
Cryptol> :safe \(x:[8]) -> 1+x
Safe
(Total Elapsed Time: 0.009s, using "Z3")

Cryptol> :safe \(x:[8]) -> 1/x
Counterexample
(\(x : [8]) -> 1 / x) 0x00 ~> ERROR
division by 0
-- Backtrace --
(Cryptol::/) called at <interactive>:7:19--7:22
<interactive>::it called at <interactive>:7:7--7:22
(Total Elapsed Time: 0.011s, using "Z3")
```

perhaps `safe` in python returns `Union[Satisfiable, Unsatisfiable]`?


I get that we're sort of doing that with the `__bool__` on `SmtUnsatResult`, but then I immediately start to wonder why `SmtSatResult` doesn't have one, what that means, and I start getting confused and feel like I need run to a whiteboard real quick and whip out some greek notation to remind myself the model-theory implications of these various terms given the question we're posing... >_<

pnwamk replied:
Also, IMHO this might be one of those cases where we're better off _not_ using class-based inheritance, since we're talking about a small finite set of possibilities and we don't get any mileage out of polymorphic OO dispatch or the like. I.e., if the main win from inheritance is they all get a `qtype` field or a particular method, then maybe we should instead use a [Protocol](https://www.python.org/dev/peps/pep-0544/) for structural subtyping like we do [here](https://github.com/GaloisInc/argo/blob/37ce3b28832ff1fad0bf18630aeed59777a36fd4/python/argo_client/interaction.py#L10-L15) in `argo-client`.

pnwamk replied:
Thinking about this a little more and chatting with some folks internally about python idioms, I think I'm personally biased towards having more specific classes (i.e., not using class names like Satisfied/Unsatisfied unless that is specifically what the user was asking for in the request, otherwise it feels like a leaking of the SMT abstraction more than is ideal IMHO) and `__bool__` seems like a nice way to make testing simple (as was already nicely implemented in this PR)

m-yac replied:
Check out the top of `synchronous.py` - there are now `Qed`, `Safe`, `Counterexample`, `Satisfiable`, and `Unsatisfiable` classes! `Qed`, `Safe`, and `Satisfiable` have `__bool__` methods which return `True`, `Counterexample` and `Unsatisfiable` have `__bool__` methods which return false, and `OfflineSmtQuery` has a bool method which throws an error (in `solver.py`).

pnwamk commented on cryptol-remote-api/python/cryptol/commands.py:
```diff
@@ -161,14 +183,62 @@ def __init__(self, connection : HasProtocolState, expr : Any) -> None:
+        return self.qtype != SmtQueryType.SAT
+
+class SmtSatResult(List[CryptolPython], SmtResult):
+    def __init__(self, qtype : SmtQueryType, cxs : List[CryptolPython]) -> None:
```
I think these parameter names would be clearer if they were a hair more verbose. I can deduce `qtype` means `query_type`, but `cxs` is making me scratch my head a little, and I can't yet decide if I'm having a bonehead moment or it really is unclear xD

m-yac replied:
Nope, it's just plain old bad naming - that was supposed to be short for "counterexample" D:

This class no longer exists, however, so this no longer needs to be fixed.

pnwamk commented on cryptol-remote-api/python/cryptol/commands.py:
```diff
@@ -161,14 +183,62 @@ def __init__(self, connection : HasProtocolState, expr : Any) -> None:
+        SmtResult.__init__(self, qtype)
+        list.__init__(self, cxs)
+
+SmtQueryResult = Union[SmtResult, OfflineSmtQuery]
```
Tell me what you think! It seems like having to worry about whether or not we got an `OfflineSmtQuery` in every instance where we call one of these functions is silly, because we can _only_ get those under certain circumstances. I.e., when the solver is _specified_ as an offline solver. Would it be nice to have `prove_offline`, `sat_offline`, and `check_offline` which _only_ returns a `OfflineSmtQuery`?

We could then be clearer about our SMT `Solver` type (I forget what it's called as I'm typing this), and have one type for the online solvers (probably called whatever it is now), and then have a _specific_ `OfflineSolver` type that delineates the available offline solvers. Then the online and offline versions of the functions could be specific about what solvers they allow, and the return types could not worry about the impossible result types! (i.e., online query will never return `OfflineSmtQuery`, etc).

Aye? Thoughts?

m-yac replied:
In the spirit of having the Python interface mirror the REPL interface, I didn't make separate methods for offline queries, I instead used `@overloads` to give the two cases (the `OnlineSolver` and `OfflineSolver` cases, as I called them) distinct types. Let me know what you think

pnwamk commented on cryptol-remote-api/python/cryptol/synchronous.py:
```diff
@@ -0,0 +1,176 @@
+from . import CryptolConnection, SmtQueryType
+
+
+def connect_sync(command : Optional[str]=None,
```
Do these functions need the `_sync` prefix, given that they're in their own module? (i.e., if you wanted to import the old and new interface, couldn't you have a prefix on the module imports to tell them apart?)

It's not a big deal, but it seems like if we move towards having the synchronous interface be the default it might be funny to have `_sync` on the methods. If anything we'd want the default interface to be smooth/streamlined and the odd one to have warts (if it has to -- i.e., if import prefixes aren't sufficient).

(Maybe I'm not seeing a reason to have the suffixes though?)

m-yac replied:
I refactored so now the tests use `cryptol.sync.connect(...)` instead of `cryptol.connect_sync(...)`. I don't know if the module trickery I did to get `sync` working as an abbreviation is very Pythonic though... see `__init__.py`

pnwamk reviewed (commented):
Love the changes. Just have some minor comments here and there, possibly bikeshedding on the interface of the SMT-related commands xD (sorry - I got confused and thought maybe a slightly simpler set of return types would help).

m-yac commented:
Looks like the CI failures have nothing to do with these RPC changes.

@pnwamk pinging you to see if you want to take another look at this to make sure I addressed your comments.

pnwamk commented on cryptol-remote-api/python/cryptol/synchronous.py:
```diff
@@ -0,0 +1,283 @@
+@dataclass
+class Counterexample:
+    """The negative result of a 'prove' or 'safe' SMT query."""
+    type : Union[Literal["predicate falsified"], Literal["safety violation"]]
```
I think we should add a note regarding this objects "falseness".

pnwamk commented on cryptol-remote-api/python/cryptol/synchronous.py:
```diff
@@ -0,0 +1,283 @@
+@dataclass
+class Unsatisfiable:
+    """The negative result of a 'sat' SMT query."""
+    def __bool__(self) -> bool:
```
I think we should add a note regarding this objects "falseness".

pnwamk commented on cryptol-remote-api/python/cryptol/synchronous.py:
```diff
@@ -0,0 +1,283 @@
+        """Check the satisfiability of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`, and return up to
+        `count` solutions.
```
We should add a note in this docstring regarding being able to boolean test `Satisfiable` and `Unsatisfiable` (i.e., that the latter is falsy).

pnwamk commented on cryptol-remote-api/python/cryptol/synchronous.py:
```diff
@@ -0,0 +1,283 @@
+        """Check the validity of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`.
+        """
```
We should add a note in this docstring regarding being able to boolean test `Qed ` and `Counterexample ` (i.e., that the latter is falsy).

pnwamk commented on cryptol-remote-api/python/cryptol/synchronous.py:
```diff
@@ -0,0 +1,283 @@
+
+    def safe(self, expr : Any, solver : Solver = Z3) -> Union[Safe, Counterexample, OfflineSmtQuery]:
+        """Check via an external SMT solver that the given term is safe for all inputs,
+        which means it cannot encounter a run-time error.
```
We should add a note in this docstring regarding being able to boolean test `Safe ` and `Counterexample ` (i.e., that the latter is falsy).

pnwamk reviewed (approved):
LGTM modulo adding just a tiny more to the docstrings of those SMT-related queries IMO.

m-yac commented:
> LGTM modulo adding just a tiny more to the docstrings of those SMT-related queries IMO.

Good catch, adding docs for this is important. What do you think of my latest commit?

pnwamk commented:
LGTM

On Wed, Sep 15, 2021 at 13:54 Matthew Yacavone ***@***.***>
wrote:

> LGTM modulo adding just a tiny more to the docstrings of those SMT-related
> queries IMO.
>
> Good catch, adding docs for this is important. What do you think of my
> latest commit?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/GaloisInc/cryptol/pull/1272#issuecomment-920368679>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA36B2RUUB52B7U5ZTYD4YDUCEB2BANCNFSM5C6GPHJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
*Andrew M. Kent, PhD*
*Research Engineer*
*Galois, Inc.*
